### PR TITLE
Refactor display modules

### PR DESCRIPTION
**Pull request by martinRenou**
This moves core display functions (for publishing output to the front-end) in a separate file.

Related issue #12223

### CODE REVIEW

**Comment by martinRenou** [comment]
Not sure I get why the CI is failing here. It looks unrelated but I'm not sure.

**Comment by martinRenou** [comment]
This PR will make quite a difference in xeus-python!
diff

**Comment by SylvainCorlay** [comment]
:+1: this is great!

**Comment by SylvainCorlay** [comment]
@Carreau would you be OK with this? It should not have any impact functionally in IPython, but would simplify the story quite a bit on the xeus side.

**Comment by Carreau** [comment]
> Not sure I get why the CI is failing here. It looks unrelated but I'm not sure.

Not your fault, new version of sphinx.

> would you be OK with this? It should not have any impact functionally in IPython, but would simplify the story quite a bit on the xeus side.

Yes, let me play with Module __getattr__, and merge a similar but backward compatible with warnings changes in 7.x. If I merge this now I'm going to forget.

**Comment by Carreau** [comment]
(also rebasing on master should fix CI)

**Comment by Carreau** [comment]
Thoughts about https://github.com/martinRenou/ipython/pull/1 ?

**Comment by martinRenou** [comment]
Thanks for your PR @Carreau. So do you prefer this approach or do you think it's nicer to move everything that is not "core" in lib?

I commented on the second approach in #12223.

**Comment by Carreau** [comment]
That is good for me if it's a clearer distinction of what was moved.

Do you want to move more objects later like DisplayObject and Co ? We can do that in subsequent PRs if needs be.

And do you feel like the `__getattr__` dance is safe enough to backport on 7.x ?

**Comment by martinRenou** [comment]
> Do you want to move more objects later like DisplayObject and Co ?

It should be just fine where it is.

> And do you feel like the __getattr__ dance is safe enough to backport on 7.x ?

I would say so. And it is definitely cleaner than what I had.

File: IPython/core/display.py [hunk 1]
```diff
@@ -5,27 +5,49 @@
...
+
+
+# ----- warn to import from IPython.display -----
+
+from warnings import warn
+
+
+def __getattr__(name):
+    if name in _deprecated_names:
+        warn(f"Importing {name} from IPython.core.display is deprecated since IPython 7.14, please import from IPython display", DeprecationWarning, stacklevel=2)
```
Comment (ivanov) [inline IPython/core/display.py:43]: I don't have a strong opinion on this, but is it even worth deprecating the location of these, if the monkey-patching that xeus-python will do will just take care of everything transparently and we can keep everyone happy without having to have them change their code?

File: IPython/core/display.py [hunk 2]
```diff
@@ -5,27 +5,49 @@
...
+    else:
+        raise AttributeError(f"module {__name__} has no attribute {name}")
 
-__all__ = ['display', 'display_pretty', 'display_html', 'display_markdown',
-'display_svg', 'display_png', 'display_jpeg', 'display_latex', 'display_json',
-'display_javascript', 'display_pdf', 'DisplayObject', 'TextDisplayObject',
-'Pretty', 'HTML', 'Markdown', 'Math', 'Latex', 'SVG', 'ProgressBar', 'JSON',
-'GeoJSON', 'Javascript', 'Image', 'clear_output', 'set_matplotlib_formats',
-'set_matplotlib_close', 'publish_display_data', 'update_display', 'DisplayHandle',
-'Video']
```
Comment (ivanov) [inline IPython/core/display.py:28]: I'd be fine with keeping this `__all__` as it was an not leaking to anyone who doesn't care too much that some of these got moved into `display_functions`. How will we get around not being able to notify someone doing `from IPython.core.display import *`?

File: IPython/display.py [hunk 3]
```diff
@@ -1,16 +1,44 @@
...
-#-----------------------------------------------------------------------------
+# -----------------------------------------------------------------------------
 
-#-----------------------------------------------------------------------------
+# -----------------------------------------------------------------------------
 # Imports
-#-----------------------------------------------------------------------------
+# -----------------------------------------------------------------------------
 
-from IPython.core.display import *
```
Comment (ivanov) [inline IPython/display.py:15]: I think it should be fine to leave this one. The point of this approach was to keep old code like this from breaking.

File: IPython/testing/iptest.py [hunk 4]
```diff
@@ -57,6 +57,7 @@
 warnings.filterwarnings('error', message='.*disable_gui.*', category=DeprecationWarning, module='.*')
 
 warnings.filterwarnings('error', message='.*ExceptionColors global is deprecated.*', category=DeprecationWarning, module='.*')
+warnings.filterwarnings('error', message='.*IPython.core.display.*', category=DeprecationWarning, module='.*')
```
Comment (ivanov) [inline IPython/testing/iptest.py:60]: and then we don't have to add more filter warnings :)

**Review by ivanov — APPROVED** [review]
Thanks, @martinRenou - this looks great. I think it'd be even better without deprecating anything. What do you think?

File: IPython/core/display.py (hunk 1 above)
Comment (Carreau) [reply IPython/core/display.py:43]: We may still remove them from here at some point, I would prefer from people to not import from core anyway.

File: IPython/core/display.py (hunk 2 above)
Comment (Carreau) [reply IPython/core/display.py:28]: the issue with keeping this all is that `import *` now import deprecated aliases (which IPython.display was doing), so was basically silencing warnings.

Comment (Carreau) [reply IPython/core/display.py:28]: I do see your point about other doing `import *`, but it should be discourage anyway. I'll think about it.

File: IPython/display.py (hunk 3 above)
Comment (Carreau) [reply IPython/display.py:15]: The problem is is was triggering the deprecation warning at import time, which then would be silenced for users, as the default is `warning:once`, hence why I'm explicit.

File: IPython/testing/iptest.py (hunk 4 above)
Comment (Carreau) [reply IPython/testing/iptest.py:60]: I want us to avoid internal imports,as if we do import a single time then we silence the warnings.

**Comment by Carreau** [comment]
Thanks @ivanov for the review. most of the changes you commented on are to make sure warnings get triggers by users. I get your point for `import*` but I also don't want trigger the warnings for people that might `import *`  and not use deprecated function. 

Xeus may do monkey patching that works, but if we can have them not do it, it's even better, and they might not be the only alternative implementation.

**Comment by martinRenou** [comment]
> Xeus may do monkey patching that works, but if we can have them not do it, it's even better, and they might not be the only alternative implementation.

It might be difficult, as of today, to find another solution than monkey-patching. Lots of Notebooks out there contain things like `from IPython.display import display, Audio`. That is the single reason why we are monkey-patching, to add support for this kind of Notebooks.

**Comment by Carreau** [comment]
What if we can detect if we are running in xeus and then import Video and Co conditionally from you or some 3rd party packages ? I do have https://pypi.org/project/disp/, and https://pypi.org/project/display/ seem to be kinda abandonned.

**Comment by martinRenou** [comment]
That was actually one idea we had. Which was to export `Video`, `Audio`.. from `core.display` and everything from `lib.display` into a separate display package that IPython and Xeus uses.

**Comment by Carreau** [comment]
Ok, i'm going to merge that with master, which will be released with 8.0 I'm not sure when. IF you want it in IPython 7.x, let me know and we can backport to 7.x.

**Pull request merged**